bouncer/src/repo/manipulator/repo_manipulator.cpp

```diff
@@ -50,54 +55,55 @@ RepoManipulator::~RepoManipulator()
 }
 
 bool RepoManipulator::connectAndAuthenticateWithAdmin(
-	std::string& errMsg,
 	const std::string& address,
 	const uint32_t& port,
 	const uint32_t& maxConnections,
 	const std::string& username,
-	const std::string& password,
-	const bool& pwDigested
+	const std::string& password
 )
 {
-	//FIXME: we should have a database manager class that will instantiate new handlers/give existing handlers
-	repo::core::handler::AbstractDatabaseHandler* handler =
-		repo::core::handler::MongoDatabaseHandler::getHandler(
-			errMsg, address, port, maxConnections,
-			repo::core::handler::MongoDatabaseHandler::getAdminDatabaseName(),
-			username, password, pwDigested);
-
-	return handler != 0;
+	repo::core::handler::MongoDatabaseHandler::ConnectionOptions options;
+	options.maxConnections = maxConnections;
+	dbHandler = repo::core::handler::MongoDatabaseHandler::getHandler(
+		address,
+		port,
+		username,
+		password,
+		options
+	);
+	try {
+		dbHandler->getCollections(ADMIN); // Test the connection to the database
+	}
+	catch (std::runtime_error)
+	{
+		return false;
```

Comment:
Agreed, I have thought some more and tweaked how this works.

`connectAndAuthenticateWithAdmin` no longer returns anything but calls the new testConnection method. The exception (if any) bubbles up to the client, where it is caught and the error code (because the mongo exceptions inherit from system_error, rather than exception) are used to infer coarsely the type of error for the return code.

The original exception object is used to print the error so those details are in the log.

While the use of error codes is an implementation detail, the alternatives are to include the mongo exception headers in the client, which is leaky, or to catch and wrap the exceptions in RepoException.

We have done the latter in places, but I am not too fond of this idea as (a) it is potentially lossy (the places we do it explicitly add context such as bson field names, but in this case it would be more likely to lose it) and (b) more importantly, if we assume we always wrap native exceptions, and miss one, it won't be handled correctly.

My feelings are that these should be very rare and fatal conditions, so its worth just letting the native ones rise and enter the log unadultered, and if we can get the right return code great, but it should be secondary to maintainability.

Reply:
I think it is better if we catch the mongo exceptions here, where we're expected it understand them, and translate them to RepoException or just print the message with additional context of what we're trying ot do then carry on with whatever we're trying to do.

Whilst this doesn't happen often when something is up and running, we do run into this a lot when we're trying to setup a deployment and helps our diagnosis. So I think I'd prefer it if we can handle it with as much interpretation as possible.

Again this could be like the the existing code, we keep the boolean check return, but just log the error here in the log with some wrapper text on what we were trying to do.